module/input/SensorFilter/tests/SensorFilter.cpp

```diff
@@ -18,4 +18,650 @@
+
+    bool failed = false;
+    std::string error_msg;
+    switch (filter.set_state(mean.getStateVec(), covariance.asDiagonal())) {
+        case Eigen::Success: break;
+        case Eigen::NumericalIssue:
+            error_msg = "Cholesky decomposition failed. The provided data did not satisfy the prerequisites.";
+            failed    = true;
+            break;
+        case Eigen::NoConvergence:
+            error_msg = "Cholesky decomposition failed. Iterative procedure did not converge.";
+            failed    = true;
+            break;
+        case Eigen::InvalidInput:
+            error_msg =
+                "Cholesky decomposition failed. The inputs are invalid, or the algorithm has been improperly called. "
+                "When assertions are enabled, such errors trigger an assert.";
+            failed = true;
+            break;
+        default:
+            error_msg = "Cholesky decomposition failed. Some other reason.";
+            failed    = true;
+            break;
+    }
+
+    if (failed) {
+        // Failed to initialise UKF
+        const double covariance_sigma_weight = 0.1 * 0.1 * 13;
+        const Eigen::Matrix<double, 13, 13> state(
+            covariance_sigma_weight * filter.getCovariance().unaryExpr([](const double& c) { return std::abs(c); }));
+        INFO(state.diagonal());
+
+        INFO(error_msg);
+        FAIL("Failed to initialise UKF. Aborting");
+    }
+
+    // Noise to be applied to gyroscope measurements
+    Eigen::Matrix3d gyroscope_noise =
+        Eigen::Vector3d(config["motion_filter"]["noise"]["measurement"]["gyroscope"].as<Expression>()).asDiagonal();
+
+    // Noise to be applied to accelerometer measurements
+    Eigen::Matrix3d accelerometer_noise =
+        Eigen::Vector3d(config["motion_filter"]["noise"]["measurement"]["accelerometer"].as<Expression>()).asDiagonal();
+    Eigen::Matrix3d accelerometer_magnitude_noise =
+        Eigen::Vector3d(config["motion_filter"]["noise"]["measurement"]["accelerometer_magnitude"].as<Expression>())
+            .asDiagonal();
+
+    // Elapsed time between each sensor read
+    constexpr double deltaT = 1.0 / 90.0;
+
+    // Set up for adding gaussian noise to the measurements
+    // Gyroscope datasheet says the MEMS device has a 0.03 dps/sqrt(Hz) noise density with a bandwidth of 50Hz
+    // Accelerometer datasheet says the MEMS device has a 220 micro-g/sqrt(Hz) noise density with a bandwidth 400Hz
+    std::random_device rd{};
+    std::mt19937 gen{rd()};
+    std::normal_distribution<> gyro_sensor_noise{0.0, 0.03 * std::sqrt(50.0) * M_PI / 180.0};
+    std::normal_distribution<> acc_sensor_noise{0.0, 22e-6 * std::sqrt(400) * module::input::G};
+
+    // Vector of quaternion errors from each timestep
+    std::vector<Eigen::Quaterniond> errors;
+    std::vector<double> angular_errors;
+
+    // Step through test data and get orientation predictions
+    for (int i = 0; i < int(quaternions.size()); ++i) {
```

Comment:
I'll preface this comment by saying that I know it's over the top, you probably already know this, and that you should probably ignore it lol

Using a dynamic sized container's size as the loop condition makes it evaluate the size every iteration. Caching the size and comparing against the cached size is a free optimisation - arguably at a very small cost of readability.

Reply:
eh